src/transport/SecureSession.cpp

```diff
@@ -78,27 +76,37 @@ CHIP_ERROR SecureSession::Init(const unsigned char * remote_public_key, const si
 void SecureSession::Reset(void)
 {
     mKeyAvailable = false;
-    mNextIV       = 0;
     memset(mKey, 0, sizeof(mKey));
 }
 
+uint64_t SecureSession::GetIV(const MessageHeader & header)
+{
+    uint64_t IV = header.GetMessageId();
+    if (header.GetSourceNodeId().HasValue())
+    {
+        uint64_t nodeID = header.GetSourceNodeId().Value();
+        IV |= nodeID & 0xffffffff00000000;
```

Comment:
XOR may be better from a security perspecitve than OR

Reply:
I discussed it with @balducci-apple offline before pushing this PR. We didn't see a benefit of XOR over OR in this scenario.

Reply:
Here is my concrete attack proposal:

I sniff node ID, say it is 1 for simplicity

Then IV(MessgeeID 1, NodeId 1) is identical with IV(messageid 0x100000001, NodeID1)

So if I sniff a 'door unlock payload' with any message id for node one, I know that sending an identical payload with message id increased by 0x100000001 will get accepted and decrypted.

OR makes me reuse IVs which is bad.

Reply:
In this case MessageID is a 32 bit value so 0x100000001 is invalid. As MessageID reaches 0xffffffff we'll need to perform a rekey operation (in practice we'll have rekeyed before this point). Once we rekey we can safely set MessageID to 0. In this case the top 4 bytes of the IV are exactly some transform of the source ID and the bottom 4 bytes are exactly the message ID.

Reply:
Incorporation of sourceNodeId in the IV value is symbolic and is not really required. We could simplify it to: IV = { 32'b0, messageId }.
But if we incorporate sourceNodeId then XOR operation would be more appropriate to use. OR operation will transform 32 most significant bits of IV to be mostly ones.

Reply:
Good catch on using XOR over OR in the calculation.